### PR TITLE
Add function 'extensionImplications'

### PR DESCRIPTION
**Pull request by shayne-fletcher**

### CODE REVIEW

**Comment by ndmitchell** [comment]
Still super worried about the complexity!

```
import qualified Data.Map as Map
import Data.Tuple.Extra

implications :: [(String, Bool, String)]
implications = undefined

implicationsMap :: Map.Map String ([String], [String])
implicationsMap = Map.fromListWith f [(a, ([c | b], [c | not b])) | (a,b,c) <- implications]
    where f (x1,y1) (x2,y2) = (x1++x2, y1++y2)
```

That's much simper. Might require you lifting into your Ext type, or also going via the Show because you don't have Ord, but simpler and very definitely not O(n^2) or O(n^3) (I really worry about a find and a nub in the middle).

**Comment by shayne-fletcher** [comment]
Very nice.

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs [hunk 1]
```diff
@@ -20,6 +22,34 @@ import StringBuffer
 import HscTypes
 import GHC.LanguageExtensions.Type
 import Data.List
+import Data.Maybe
+import qualified Data.Map as Map
+
+-- | Parse a GHC extension
+readExtension :: String -> Maybe Extension
+readExtension x = Map.lookup x exts
```
Comment (ndmitchell) [inline src/Language/Haskell/GhclibParserEx/DynFlags.hs]:
```suggestion
readExtension = \x -> Map.lookup x exts
```

The difference is the difference between computing the `exts` table each and every time, or doing it once.

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs [hunk 2]
```diff
@@ -20,6 +22,34 @@ import StringBuffer
...
+#else
+          | x <- [Cpp .. StandaloneKindSignatures]]
+#endif
+
+-- | Implicitly enabled/disabled extensions.
+extensionImplications :: [(Extension, ([Extension], [Extension]))]
+extensionImplications = Map.foldlWithKey f [] implicationsMap
+  where
+    f acc e (es, ds) =
+      (fromJust (readExtension e), (es, ds)) : acc
```
Comment (ndmitchell) [inline src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Why not just use `map f . Map.toList`?

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs [hunk 3]
```diff
@@ -20,6 +22,34 @@ import StringBuffer
...
+
+    implications :: [(String, Bool, Extension)]
+    implications = foldl' f [] impliedXFlags
+      where f acc (e, flag, e') = (show e, flag == turnOn, e') : acc
```
Comment (ndmitchell) [inline src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Why not just use `map`?

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs (hunk 1 above)
Comment (shayne-fletcher) [reply src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Excellent.

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs (hunk 2 above)
Comment (shayne-fletcher) [reply src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Because ... stupid 😁

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs (hunk 3 above)
Comment (shayne-fletcher) [reply src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Yep.

**Comment by shayne-fletcher** [comment]
Thanks for the help and suggestions. Think this is about done. Want a last look?

File: src/Language/Haskell/GhclibParserEx/DynFlags.hs [hunk 4]
```diff
@@ -20,6 +22,27 @@ import StringBuffer
...
+extensionImplications = map f $ Map.toList implicationsMap
+  where
+    f (e, ps) = (fromJust (readExtension e), ps)
+
+    implications :: [(String, Bool, Extension)]
+    implications = map f impliedXFlags
+      where f (e, flag, e') = (show e, flag == turnOn, e')
+
+    implicationsMap :: Map.Map String ([Extension], [Extension])
+    implicationsMap = Map.fromListWith f [(a, ([c | b], [c | not b])) | (a, b, c) <- implications]
```
Comment (ndmitchell) [inline src/Language/Haskell/GhclibParserEx/DynFlags.hs]: I'd probably "fuse" the impliciations into this function, e.g.

```suggestion
    implicationsMap = Map.fromListWith f [(show a, ([c | b], [c | not b])) | (a, bb, c) <- implications, let b = bb == turnOn]
```

Comment (shayne-fletcher) [reply src/Language/Haskell/GhclibParserEx/DynFlags.hs]: Ok!

**Comment by shayne-fletcher** [comment]
I think this idea was why you suggested `Data.Tuple.Extra`?
```
    implicationsMap = Map.fromListWith (<>)
      [(show a, ([c | b], [c | not b]))
        | (a, flag, c) <- impliedXFlags, let b = flag == turnOn]
```
Seems to test out OK (use of `<>` -- not sure why/how it's in scope without `Data.Tuple.Extra`). Too cryptic?

**Comment by ndmitchell** [comment]
I was originally doing to do something like `both (++)` from Data.Tuple.Extra, but that's not quite right. Using `<>` seems fine.

**Pull request merged**